### PR TITLE
added support for is_valid for nonlinear constraints

### PR DESCRIPTION
**Pull request by hdavid16**
as suggested by @odow in 
https://github.com/jump-dev/JuMP.jl/issues/2355#issuecomment-840068136

### CODE REVIEW

File: src/constraints.jl [hunk 1]
```diff
@@ -305,6 +305,22 @@ function is_valid(model::Model, con_ref::ConstraintRef{<:AbstractModel})
...
 
+"""
+    is_valid(model::Model,
+             c::ConstraintRef{Model,NonlinearConstraintIndex})
+
+Return `true` if `constraint_ref` refers to a valid nonlinear constraint in
+    `model`.
+"""
+function is_valid(model::Model,
+                  c::ConstraintRef{Model,NonlinearConstraintIndex})
```
Comment (odow) [inline src/constraints.jl:316]: This is going to fail the format check. Wait for CI to fail and follow the diff, or use https://jump.dev/JuMP.jl/stable/developers/style/#JuliaFormatter.

The formatter doesn't format docstrings, but you should update it to match, and fix the indent on line 313.

**Review by odow — COMMENTED** [review]
Can you add some tests please?

Somewhere near here:
https://github.com/jump-dev/JuMP.jl/blob/80626541cc19005d1cb48958531d2ec71e8b7bf7/test/nlp.jl#L1044-L1051

File: src/constraints.jl [hunk 2]
```diff
@@ -305,6 +305,23 @@ function is_valid(model::Model, con_ref::ConstraintRef{<:AbstractModel})
...
+function is_valid(
+    model::Model,
+    c::ConstraintRef{Model,NonlinearConstraintIndex},
+)
+    if model !== c.model
+        return false
+    end
+    _init_NLP(model)
+    return 1 <= c.index.value <= num_nl_constraints(model)
+end
```
Comment (odow) [inline src/constraints.jl:323]: Ah. I missed this on first glance. It needs to go in the `src/nlp.jl` file after `NonlinearConstraintIndex` is defined.

Comment (hdavid16) [reply src/constraints.jl:323]: yeah, I'm moving it over there. Noticed it after precompilation errors came up

File: src/constraints.jl (hunk 1 above)
Comment (hdavid16) [reply src/constraints.jl]: done

**Comment by hdavid16** [comment]
added tests at the end of the nlp.jl tests file

**Review by odow — APPROVED** [review]
Looks good, assuming tests pass.

**Pull request merged**